### PR TITLE
Large import

### PR DESCRIPTION
**Pull request by benjiec**
Reduce large chunks to smaller chunks, during import.

### CODE REVIEW

File: src/edge/importer.py [hunk 1]
```diff
@@ -107,9 +107,28 @@ def build_fragment(self):
...
+                    divided_chunks.append(min(big_cs, cs_limit))
+                    big_cs -= cs_limit
+                new_chunk_sizes.extend(divided_chunks)
+        chunk_sizes = new_chunk_sizes
+
         for sz in chunk_sizes:
+            t0 = time.time()
             prev = new_fragment._append_to_fragment(prev, flen, self.__sequence[flen:flen + sz])
             flen += sz
+            nn = new_fragment.fragment_chunk_location_set.count()
```
Comment (yaoyuyang) [inline src/edge/importer.py:128]: what does nn mean? can we use a more descriptive variable name?

File: src/edge/importer.py [hunk 2]
```diff
@@ -107,9 +107,28 @@ def build_fragment(self):
         prev = None
         flen = 0
         seqlen = len(self.__sequence)
+
+        cs_limit = 1000000
```
Comment (yaoyuyang) [inline src/edge/importer.py:111]: adding a comment here to briefly describe the rationale of breaking up the big chunk sizes into smaller sizes?

File: src/edge/importer.py [hunk 3]
```diff
@@ -107,9 +107,28 @@ def build_fragment(self):
...
+        new_chunk_sizes = []
+        for big_cs in chunk_sizes:
```
Comment (yaoyuyang) [inline src/edge/importer.py:113]: rename big_cs as big_chunk_size? it will be easier to read.

File: src/edge/importer.py [hunk 4]
```diff
@@ -122,9 +141,12 @@ def annotate(self, fragment):
                                                          .filter(fragment=fragment)}
 
         for feature in self.__features:
+            t0 = time.time()
             f_start, f_end, f_name, f_type, f_strand, f_qualifiers = feature
             # print('  %s %s: %s-%s %s' % (f_type, f_name, f_start, f_end, f_strand))
             self._annotate_feature(fragment, f_start, f_end, f_name, f_type, f_strand, f_qualifiers)
+            print('annotate feature: %.4f\r' % (time.time() - t0,), end="")
```
Comment (yaoyuyang) [inline src/edge/importer.py:148]: Ideally, we should convert all these print statements into logger but let's leave that in the future refactoring.

File: src/edge/importer.py [hunk 5]
```diff
@@ -142,10 +169,13 @@ def _annotate_feature(self, fragment, first_base1, last_base1, name, type, stran
                 (name, first_base1, last_base1))
 
         annotation_start = self.__fclocs[first_base1]
-        if last_base1 != len(self.__sequence):
```
Comment (yaoyuyang) [inline src/edge/importer.py:145]: Was this a bug before?

File: src/edge/importer.py (hunk 2 above)
Comment (yaoyuyang) [reply src/edge/importer.py:117]: Thanks! This is super clear now!

File: src/edge/importer.py [hunk 6]
```diff
@@ -142,10 +169,13 @@ def _annotate_feature(self, fragment, first_base1, last_base1, name, type, stran
...
+        if last_base1 < len(self.__sequence):
             annotation_end = self.__fclocs[last_base1 + 1]
-        else:
+        elif last_base1 == len(self.__sequence):
             annotation_end = self.__fclocs[1]
+        else:
+            # bad annotation, beyond end of sequence
```
Comment (yaoyuyang) [inline src/edge/importer.py:177]: Do we need to throw any warnings here or does it not matter at all?

File: src/edge/importer.py [hunk 7]
```diff
@@ -107,9 +110,30 @@ def build_fragment(self):
         prev = None
         flen = 0
         seqlen = len(self.__sequence)
+
+	# divide chunks bigger than a certain threshold to smaller chunks, to
+	# allow insertion of sequence into database. e.g. MySQL has a packet
+	# size that prevents chunks that are too large from being inserted.
+        cs_limit = 1000000
```
Comment (yaoyuyang) [inline src/edge/importer.py:117]: Thanks for updating the variable names, super clear now. Do you mind just change this cs_limit to chunk_size_limit to be consistent? Thanks!

File: src/edge/importer.py [hunk 8]
```diff
@@ -107,9 +110,30 @@ def build_fragment(self):
...
+                new_chunk_sizes.append(original_chunk_size)
+            else:
+                divided_chunks = []
+                while original_chunk_size > 0:
+                    divided_chunks.append(min(original_chunk_size, cs_limit))
+                    original_chunk_size -= cs_limit
+                new_chunk_sizes.extend(divided_chunks)
+        chunk_sizes = new_chunk_sizes
+
         for sz in chunk_sizes:
```
Comment (yaoyuyang) [inline src/edge/importer.py:130]: Do you mind updating this sz to be something like chunk_size while we are here? Thanks!

File: src/edge/importer.py (hunk 5 above)
Comment (benjiec) [reply src/edge/importer.py:145]: Yes it was

File: src/edge/importer.py (hunk 6 above)
Comment (benjiec) [reply src/edge/importer.py:177]: I don't know how to throw a warning to the user.

Comment (yaoyuyang) [reply src/edge/importer.py:177]: It's fine to keep it this way for now. I think in the future if we have the bandwidth, we might want to introduce logger so we can have different levels of logging where the other applications can catch. for example, logger.info, logger.warning, etc.

File: src/edge/importer.py (hunk 5 above)
Comment (yaoyuyang) [reply src/edge/importer.py:145]: Nice you found and fixed it!

**Pull request merged**